compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -217,6 +240,189 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
       if tvar == null then NoType
       else tvar
 
+// ------------- Type parameter dependencies ----------------------------------------
+
+  private type ReverseDeps = SimpleIdentityMap[TypeParamRef, SimpleIdentitySet[TypeParamRef]]
+
+  /** A map that associates type parameters of this constraint with all other type
+   *  parameters that refer to them in their bounds covariantly, such that, if the
+   *  type parameter is instantiated to a larger type, the constraint would be narrowed
+   *  (i.e. solution set changes other than simply being made larger).
+   */
+  private var coDeps: ReverseDeps = SimpleIdentityMap.empty
```

Comment:
I only just realised that this made OrderingConstraint into a mutable data structure (previously it was immutable).

Reply:
It's actually not mutable. `coDeps` and `contraDeps` are views that are maintained incrementally.

Reply:
That would only make it not destructive.  And even so, there is a `dropDeps`, which does remove from them, if I'm not mistaken.

Reply:
I am not sure what you mean by destructive vs mutable. Are LazyLists mutable? In my book they are. And, `dropDeps` is not exported in `Constraint` so it is morally private (we should probably make it private to be clear about it)

Reply:
Ah, I understand better now.  You're right.

Even though it looks very different, I think I prefer the approach that's taken with the original components of OrderingConstraint, i.e. in ConstraintLens update, where depending on the previous constraint (which is the external, public reference) it was either mutated in place or copied on write, and then always returned the updated reference.  And that logic was well separate from the rest of the domain logic.  With these new variables, the updates just happen at some point of the logic, mutating the vars in place.